exonum/src/api/ext.rs

```diff
@@ -0,0 +1,542 @@
+//!     }
+//! }
+//! # fn main() { }
+//! ```
+
+use serde::Serialize;
+use serde::de::DeserializeOwned;
+use serde_json::{self, Value};
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::{fmt, io};
+
+use blockchain::{Blockchain, SendError, Transaction};
+use crypto::{Hash, PublicKey};
+use messages::RawMessage;
+use storage::Snapshot;
+
+/// The specification for the "standard" transaction sink.
+///
+/// Transaction sinks can be added to service APIs using the [`set_transactions`]
+/// method in `ServiceApi`; see it for more details.
+///
+/// [`add_transactions`]: struct.ServiceApi.html#method.set_transactions
+pub const TRANSACTIONS: Spec = Spec { id: "transactions" };
+
+/// API-related errors.
+#[derive(Debug, Fail)]
+pub enum ApiError {
+    /// Call to endpoint with unknown identifier.
+    #[fail(display = "Unknown endpoint ID: {}", _0)]
+    UnknownId(String),
+
+    // TODO: split `serde::Error` / others?
+    /// A request is malformed or otherwise cannot be processed.
+    #[fail(display = "Bad request: {}", _0)]
+    BadRequest(Box<::std::error::Error + Send + Sync>),
+
+    /// Requested resource is not found.
+    #[fail(display = "Not found")]
+    NotFound,
+
+    /// Transaction verification has failed.
+    #[fail(display = "Transaction not verified: {:?}", _0)]
+    VerificationFail(Box<Transaction>),
+
+    /// A transaction processed by API has failed to be sent.
+    #[fail(display = "Failed to send transaction: {}", _0)]
+    TransactionNotSent(
+        #[cause]
+        io::Error
+    ),
+
+    /// Generic server-side error.
+    #[fail(display = "Internal server error: {}", _0)]
+    InternalError(Box<::std::error::Error + Send + Sync>),
+}
+
+impl From<SendError> for ApiError {
+    fn from(err: SendError) -> ApiError {
+        use self::SendError::*;
+
+        match err {
+            VerificationFail(tx) => ApiError::VerificationFail(tx),
+            Io(e) => ApiError::TransactionNotSent(e),
+        }
+    }
+}
+
+/// Alias for the result type used within the `api` module.
+pub type ApiResult<T> = Result<T, ApiError>;
+
+/// Specification of an endpoint.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Spec {
+    /// Endpoint identifier.
+    pub id: &'static str,
+}
+
+/// Context supplied to read-only endpoints.
+#[derive(Debug)]
+pub struct ReadContext<'a> {
+    blockchain: &'a Blockchain,
+}
+
+impl<'a> ReadContext<'a> {
+    fn new(blockchain: &'a Blockchain) -> Self {
+        ReadContext { blockchain }
+    }
+
+    /// Gets a snapshot of the current blockchain state.
+    pub fn snapshot(&self) -> Box<Snapshot> {
+        self.blockchain.snapshot()
+    }
+
+    /// Gets the public key used for signing operations by services.
+    pub fn service_public_key(&self) -> &PublicKey {
+        self.blockchain.service_public_key()
+    }
+}
+
+/// Context supplied to mutating endpoints.
+#[derive(Debug)]
+pub struct MutContext<'a> {
```

Comment:
It sounds a little bit inconsistent with the `ReadContext`.

Reply:
Yeah, I don't quite like the naming of context types. Maybe, `EndpointContext` / `MutEndpointContext`? Or use shorter names, assuming that devs will qualify if necessary (e.g., `api::Context` / `api::MutContext`)?